central/sensor/service/pipeline/imageintegrations/pipeline.go

```diff
@@ -226,4 +230,75 @@ func (s *pipelineImpl) Run(ctx context.Context, clusterID string, msg *central.M
 	return nil
 }
 
+func (s *pipelineImpl) runRemove(ctx context.Context, id string) error {
+	if err := s.integrationManager.Remove(id); err != nil {
+		return errors.Wrap(err, "removing integration from manager")
+	}
+	return s.datastore.RemoveImageIntegration(ctx, id)
+}
+
+// Run runs the pipeline template on the input and returns the output.
+// Action is currently always update.
+func (s *pipelineImpl) Run(ctx context.Context, clusterID string, msg *central.MsgFromSensor, _ common.MessageInjector) error {
+	// Ignore autogenerated registries if they are disabled
+	if autogeneratedRegistriesDisabled {
+		return nil
+	}
+	defer countMetrics.IncrementResourceProcessedCounter(pipeline.ActionToOperation(msg.GetEvent().GetAction()), metrics.ImageIntegration)
+
+	if msg.GetEvent().GetAction() == central.ResourceAction_REMOVE_RESOURCE {
+		return s.runRemove(ctx, msg.GetEvent().GetImageIntegration().GetId())
+	}
+
+	// Extract the cluster name.
+	clusterName, exists, err := s.clusterDatastore.GetClusterName(ctx, clusterID)
+	if err != nil {
+		return errors.Wrapf(err, "error getting cluster name for cluster ID: %s", clusterID)
+	}
+	if !exists {
+		return errox.NotFound.Newf("cluster with id %q does not exist", clusterID)
+	}
+
+	// Set up the integration and update its parameters.
+	imageIntegration := msg.GetEvent().GetImageIntegration()
+	imageIntegration.ClusterId = clusterID
+
+	description, matches, err := setUpIntegrationParams(ctx, imageIntegration)
+	if err != nil {
+		return errors.Wrapf(err, "setting up integration params for %q", imageIntegration.GetId())
+	}
+	source := imageIntegration.GetSource()
+	if source == nil {
+		return s.legacyRun(ctx, imageIntegration, matches)
+	}
+	imageIntegration.Name = fmt.Sprintf("Autogenerated %s for cluster %s from %s/%s",
+		description, clusterName, source.GetNamespace(), source.GetImagePullSecretName())
+
+	requiresShortCircuit, err := s.upsertImageIntegration(ctx, imageIntegration)
+	if err != nil {
+		return err
+	}
+
+	if requiresShortCircuit {
+		s.enrichAndDetectLoop.ShortCircuit()
+	}
+
+	return nil
+}
+
 func (s *pipelineImpl) OnFinish(_ string) {}
+
+func (s *pipelineImpl) upsertImageIntegration(ctx context.Context, imageIntegration *storage.ImageIntegration) (bool, error) {
+	_, exists, err := s.datastore.GetImageIntegration(ctx, imageIntegration.GetId())
+	if err != nil {
+		return exists, errors.Wrapf(err, "retrieving image integration %q", imageIntegration.GetId())
+	}
+
+	if _, err := s.datastore.AddImageIntegration(ctx, imageIntegration); err != nil {
+		return exists, errors.Wrapf(err, "adding image integration %q", imageIntegration.GetId())
+	}
+	if err := s.integrationManager.Upsert(imageIntegration); err != nil {
+		return exists, errors.Wrapf(err, "notifying of update for image integration %q", imageIntegration.GetId())
+	}
+	return exists, nil
```

Comment:
Didn't you now flip when to do a short-circuit? Why would we _not_ want to short circuit upon creation of an image integration?

Reply:
Oh well, yup.